### PR TITLE
(#2710) Update repository documentation around testing

### LINKED ISSUE #2710: Update Readme/Contributing doc to point to new testing doc
When we added the Pester tests (#2597), we added a `TESTING.md` file. That file isn't really referenced from anywhere.

This issue is to track adding references to that from the Readme and Contributing documents, and possibly consolidating some of the content from those files into the Testing file.

### LINKED ISSUE #2762: Vagrant environment for running tests should have a way to allow for testing against released versions of Chocolatey
## Is Your Feature Request Related To A Problem? Please describe.

When fixes for issues in Chocolatey are added, we usually add a Pester test to ensure we don't regress on the bug. This is good, however if we forget to gate those tests against the next version of Chocolatey, they will cause Test Kitchen tests to fail when run against the current stable version.

## Describe The Solution. Why is it needed?

It would be good to have a way to test against the stable version of Chocolatey inside of the Vagrant environment so we could identify tests that fail before merging them in and seeing failures in other test scenarios.

## Additional Context.

#2726 introduced a fix for side by side installs. The tests for this are broken on stable chocolatey and missed being gated in the tests.

## Related Issues

N/A

### PR DESCRIPTION
**Pull request by corbob**
## Description Of Changes

* Update Contributing.md to move testing section into Testing.md.
* Update Testing.md to include some guidelines to help reduce flaky tests.

## Motivation and Context

Consolidate testing information into a single location. Improve communication of test guidelines.

## Testing

N/A

## Change Types Made

* [ ] Bug fix (non-breaking change)
* [ ] Feature / Enhancement (non-breaking change)
* [ ] Breaking change (fix or feature that could cause existing functionality to change)
* [x] Documentation change.
* [ ] PowerShell code changes.

## Related Issue

Fixes #2710

## Change Checklist

* [ ] Requires a change to the documentation
* [x] Documentation has been updated
* [ ] Tests to cover my changes, have been added
* [ ] All new and existing tests passed.
* [ ] PowerShell v2 compatibility checked.

### CODE REVIEW

**Comment by gep13** [comment]
@corbob just wanted to check that you intended to close this PR, since the two linked issues are still open.

**Comment by corbob** [comment]
I was doing cleanup of my repositories last night, and either missed that this branch had a PR, or GitKraken didn't show it. In either case it was unintentional to close. I have closed one of the linked issues, as it's not actually needed due to process changes. I'll restore this branch to hopefully get back to updating the testing instructions.

**Review by gep13 — CHANGES_REQUESTED** [review]
Some small bits of feedback

File: TESTING.md [hunk 1]
```diff
@@ -1,11 +1,118 @@
...
+The exception to this rule is tests that require another product of a specific version.
+For example: a test that requires Chocolatey Licensed Extension version 6.1 or greater, but is a part of the Chocolatey CLI tests.
+
+### Tests Should Not Contain Ambiguous Statements.
+
+Whenever possible, it should be possible to identify exactly when a line of code will be executed during a test run.
+
+Currently all Pester tests are run against Pester 5.x.
+As such, code that is within a `.Tests.ps1` file, but not within any Pester specific script block will generally be executed during the `BeforeDiscovery` phase.
+However, this execution pattern is not necessarily guaranteed and
```
[the remark this answers is not in the document]
Comment (corbob) [reply TESTING.md]: Yeah, it looks like I got distracted while writing this and didn't finish it. I've revamped this section.

File: TESTING.md [hunk 2]
```diff
@@ -1,11 +1,118 @@
...
+In those instances, a version gate would be used.
+It is expected that the tests being run will be for the version being built from the same point in the repository.
+As such, we should not be skipping tests due to a version mismatch.
+
+The exception to this rule is tests that require another product of a specific version.
+For example: a test that requires Chocolatey Licensed Extension version 6.1 or greater, but is a part of the Chocolatey CLI tests.
+
+### Tests Should Not Contain Ambiguous Statements.
+
+Whenever possible, it should be possible to identify exactly when a line of code will be executed during a test run.
```
[the remark this answers is not in the document]
Comment (corbob) [reply TESTING.md]: As with the other comment about it ending abruptly: Yeah, it looks like I got distracted while writing this and didn't finish it. I've revamped this section.

File: TESTING.md [hunk 3]
```diff
@@ -1,11 +1,118 @@
...
+Below is a list of words.
+
+### A Test Or Group Of Tests Should Be Self Contained.
+
+Everything needed for a test to pass consistently should be within a single test structure.
+Whenever possible, you should be able to select any test case and run it without failure.
+
+### Tests Should Not Depend Upon The Order That They Are Executed.
+
+Coinciding with that tests should be self container, they should also not depend on other tests running in a specific order.
```
[the remark this answers is not in the document]
Comment (corbob) [reply TESTING.md]: Reading over this section again, I didn't like the use of the `coinciding` part. It felt like using big words just for the sake of using big words.

File: TESTING.md [hunk 4]
```diff
@@ -1,11 +1,118 @@
...
+  - **Physical Integration Testing** - This is where you are testing the application with other components such as config files.
+  - **Blackbox Integration Testing / End to End Testing** - This is where you are testing inputs and outputs of the system.
+- **Version Gate** - A check that is performed to ensure certain versions are in use before performing tests.
+- **Test Structure** - All components that make up a single test. In Pester this would be a `Describe` or `Context` block. In Nunit this would be the containing class.
+
+## Writing Tests
+
+The purpose of the tests we write for Chocolatey products is to ensure that we do not regress on issues that have been fixed.
+Part of ensuring that is to do the best we can to reduce test flakiness.
+Below is a list of words.
```
[the remark this answers is not in the document]
Comment (corbob) [reply TESTING.md]: This line is practically perfect in every way! (or yeah, doesn't make a lick of sense 😂 )

**Review by gep13 — APPROVED** [review]
LGTM!

**Comment by gep13** [comment]
@corbob thanks for getting this updated!

**Pull request merged**